### ISSUE TITLE
Is normal that memory allocations happens when filling?

### ISSUE THREAD

**Issue by peremato** [Moelf/FHist.jl#81]
The package is great, but while chasing  memory allocations in my application I noticed that it generate them.
See the following test case:
```
using FHist
data = randn(100);
h = Hist1D(; bins=-10:1:10)

@time for d in data
              push!(h,d)
              end
  0.000027 seconds (200 allocations: 4.688 KiB)
```
2 allocations per fill.

**Comment by Moelf** [comment]
omfg

https://github.com/Moelf/FHist.jl/blob/623685c69e4e235f8b9c57abadf9e3867c7d9f9b/src/FHist.jl#L29

this needs to be `Base.RefValue{Int}` otherwise it's an Abstract type. Thank you for pointing this out!

```julia
julia> const data = rand(Int, 100);

julia> const h = Hist1D(; bins=-10:1:10);

julia> function f(h, data)
           for d in data
               push!(h,d)
           end
           nothing
       end

julia> @ballocated f(h, data)
0
```

**Comment by peremato** [comment]
Thanks. I do have a related equation. I have struct with histograms that are shared between several parts of the application. See for [example](https://github.com/peremato/Geant4.jl/blob/af9ebfa237e1f2daed808a2931d0df3186b899b2/examples/TestEm3/TestEm3.jl#L41). I use the abstract type `Hist1D` and whenever I need to use it there is an allocation (I guess because the real type is not known). Is there a way to have an efficient struct with histograms.

**Comment by Moelf** [comment]
Right, if you just use `Hist1D` it's under-parametrized. In this use case, I think you can have
```julia
mutable struct TestEm3SimData{HT} <: G4JLSimulationData\
...
...
fTrackLengthChHistos::Vector{Hist1D}
```

**Comment by peremato** [comment]
Yes, I see that this would work. But if I have to have a parameter for each type of histogram will be a bit unmanageable.

**Comment by Moelf** [comment]
if you only use a specific type, you can make an alias for it and hard-code it?

```julia
const _HT = Hist1D{Float64, Tuple{StepRange{Int64, Int64}}}
```

basically one of the parameter track bin content type, the other track the type of "edges"

**Comment by Moelf** [comment]
@peremato feel free to keep talking here btw, or start a discussion, closed issue doesn't mean we can't keep discussion usage!

**Comment by peremato** [comment]
Typically are users (physicists) creating these kind of structures with counters, histograms, etc. So, if the user needs to know the specific type `Hist1D{Float64, Tuple{StepRange{Int64, Int64}}}` or even worst `Hist1D{Float64, Tuple{StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}}` would not be very user-friendly. What about defining some alias in the FHist package itself? For instance `Hist1Dd` or `Hist2Df`.

**Comment by Moelf** [comment]
yeah but in Geant4 you're hard coding `String` and `Float64` instead of parametrizing `<:AbstractString` or `<:AbstractFloat` so it's the same for Histogram?

>For instance Hist1Dd or Hist2Df

doesn't quite solve the problem because the edges type is unknown. Users often don't think about edges:
```julia
0.1:0.3
vs.
[0.1, 0.2, 0.3]
```

If user needs it they just have to type it out or make their struct parameterized

**Comment by peremato** [comment]
I think is OK to parametrize with the float type, but the edges looks to me an implementation detail. 
You could create histograms with (as in ROOT) `Hist1D{T<:Real}(min::T, max::T, nbins::Int64)`. In this case the type would be `Hist1D{Float64}` which is very reasonable.

**Comment by peremato** [comment]
I am from the 'old' school, that the implementation should do the best without showing too many details to the user. If the bins are not evenly spaced you can provide a constructor accepting a Vector or AbstractVector with the edges. The implementation then used a Union or equivalent to map both cases at once.

**Comment by Moelf** [comment]
>If the bins are not evenly spaced you can provide a constructor accepting a Vector or AbstractVector with the edges. 

that's exactly what we do. And we can't union over all possible `<:AbstractVector` types, so we have to parameterized it:

**Comment by peremato** [comment]
In reality there are only two:
```
typeof(min::T : bin::T : max::T)
Vector{T}
```
where T is a Real

**Comment by Moelf** [comment]
yeah but T can be many things so there are 2xN?

**Comment by peremato** [comment]
I was assuming that an `Hist1D{Float64}` the T in the bin edges is also Float64. I know this is not exactly correct but for any practical purpose is OK. So what you do is to define the range as `T(min):(max-min)/nbins:T(max)`

**Comment by Moelf** [comment]
sometimes people have bin content as Integer, and they go higher than the range `Float64` can represent exactly, meanwhile, they have Float64 bin edges
